src/coreclr/nativeaot/Runtime/riscv64/WriteBarriers.S

```diff
@@ -254,28 +244,33 @@ LEAF_END RhpCheckedAssignRef, _TEXT
 // reside on the managed heap.
 //
 // On entry:
-//  t5 : the destination address (LHS of the assignment).
-//  t6 : the object reference (RHS of the assignment).
+//  t3 : the destination address (LHS of the assignment).
+//  t4 : the object reference (RHS of the assignment).
 //
 // On exit:
-//  t3, t4 : trashed
-//  t6 : incremented by 8
+//  t2, t6 : trashed
+//  t3 : incremented by 8
 LEAF_ENTRY RhpAssignRefRiscV64, _TEXT
 
     ALTERNATE_ENTRY RhpAssignRefAVLocation
-        sd    t6, 0(t5)
+        mv    t5, a1
+        mv    a1, t3
+        sd    t4, 0(t3)
 
-        INSERT_UNCHECKED_WRITE_BARRIER_CORE t5, t6
+        INSERT_UNCHECKED_WRITE_BARRIER_CORE t3, t4
 
-        addi  t5, t5, 8
-        jalr  ra
+        addi  t3, t3, 8
+
+        // Jump to the return address in t5 (previously a1) for the tailcall.
```

Comment:
```suggestion
        // TODO-RISCV64: Jump to the return address in t5 (previously a1) for the tailcall. This should be revert to ret instead of jr once JIT is fixed
```